### PR TITLE
ninja: Fix quoting newlines and $

### LINKED ISSUE #1643: unparsable ninja.build generated from run_command with a multi-line string
```meson
project('meson7' ,'c')
run_target(
  'update-rules',
  command : ['sh', '-c',
	     '
python3 make-man-rules.py --meson'])
```

```
ninja: error: build.ninja:64: expected '=', got lexing error
python3$ make-man-rules.py$ --meson'
       ^ near here
WARNING: Could not create compilation database.
```

### PR DESCRIPTION
**Pull request by QuLogic**
The first commit pulled out of #1434; should fix #1643.

### CODE REVIEW

**Comment by QuLogic** [comment]
Need to write a test, probably.

**Comment by nirbheek** [comment]
Test failure seems to be relevant?

**Comment by QuLogic** [comment]
Yep, seems it needed a minor correction.

**Comment by QuLogic** [comment]
Rebased and a test added.

File: test cases/common/149 special characters/meson.build [hunk 1]
```diff
@@ -0,0 +1,13 @@
+project('ninja special characters' ,'c')
+
+python = import('python3').find_python()
+
+ret = run_command(python, ['check_quoting.py', 'newline=
+', 'dollar=$', 'colon=:', 'space= '])
```
Comment (nirbheek) [inline test cases/common/149 special characters/meson.build]: We should also check that multiple newlines and spaces are also preserved as-is.

Comment (QuLogic) [inline test cases/common/149 special characters/meson.build]: Added.

File: test cases/common/149 special characters/meson.build [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ret = run_command(python,
+  ['check_quoting.py',
+   'newline=
+',
+   'dollar=$',
+   'colon=:',
+   'space= ',
+   'multi= :$
+ :$
+'])
```
Comment (nirbheek) [inline test cases/common/149 special characters/meson.build]: This tests a string with two newlines separated by non-whitespace characters. My concern was that newlines and whitespaces might get 'squashed'. Could you check if two newlines in a row also works? Just adding another newline at the end of this should work, I think.

Comment (QuLogic) [inline test cases/common/149 special characters/meson.build]: I doubled some stuff and mixed it around a bit; should definitely ensure there's no squashing going on.

Comment (nirbheek) [inline test cases/common/149 special characters/meson.build]: Thanks!

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by QuLogic** [comment]
Probably doesn't need any documentation changes.

**Comment by jpakkane** [comment]
This test does not actually check that Ninja quoting works. Neither run_command nor the postconf scripts are launched via Ninja. This is easy to check, because if you convert the run_command to a run_target and then do `ninja <run_target_name` the output is this:

    Traceback (most recent call last):
      File "/Users/jpakkane/quoting/test cases/common/149 special characters/check_quoting.py", line 16, in <module>
        raise RuntimeError('{!r} is {!r} but should be {!r}'.format(name, value, expected[name]))
    RuntimeError: 'newline' is '' but should be '\n'
    FAILED: checker

**Comment by jpakkane** [comment]
Similarly you should use triple quotes to contain a newline. Having an single quote split over multiple lines should be a syntax error, just like it is in Python. I don't know why that even works.

**Comment by QuLogic** [comment]
Yes, you're right. I based that on a test that didn't need to run through Ninja; I hoped the post stuff would do it but now I remember it runs through a wrapper.

I think, after further investigation, it's not possible to pass a literal newline through ninja, as the `$\n` construct simply means a line continuation. We should probably try to catch this case higher up if it can happen and redirect those cases through the `exe` wrapper.

**Comment by QuLogic** [comment]
I modified this to use serialization if a `CustomTarget` command-line includes newlines. I don't know of other cases, but I put in an exception asking to report it here.

**Comment by jpakkane** [comment]
Interestingly this one just got more important. We are using rpath so binaries can be executed from the build dir. Currently it uses absolute paths. This breaks reproducible builds because different build dirs have different lengths, and thus the output is different. We'll need to move to relative rpaths, which means passing `$ORIGIN` through a Ninja invocation.

**Comment by QuLogic** [comment]
That should work here; at least, I tried changing `linuxlike/7 library versions` to use `$ORIGIN` and it is correctly escaped with this PR.

**Comment by jpakkane** [comment]
I would call this release notes worthy. Not claiming that quoting works perfectly but something along the lines of "better quoting of special characters in backend command invocations".

**Comment by QuLogic** [comment]
Hmm, it's not super clear to me what's wrong on AppVeyor; it doesn't affect every build with the ninja backend, and there are no obvious error messages.

**Comment by QuLogic** [comment]
Never mind, I found it; was looking at the log at the end only when the error was within the execution part.

**Pull request merged**